File: chat-uikit/src/main/java/io/agora/chat/uikit/widget/chatrow/EaseChatRowVideo.java
package io.agora.chat.uikit.widget.chatrow;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.drawable.Drawable;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.ContextCompat;

import io.agora.chat.ChatMessage;
import io.agora.chat.FileMessageBody;
import io.agora.chat.VideoMessageBody;
import io.agora.chat.uikit.R;
import io.agora.chat.uikit.chat.model.EaseChatItemStyleHelper;
import io.agora.chat.uikit.utils.EaseDateUtils;
import io.agora.chat.uikit.utils.EaseImageUtils;
import io.agora.chat.uikit.utils.EaseUtils;
import io.agora.util.EMLog;
import io.agora.util.TextFormater;


public class EaseChatRowVideo extends EaseChatRowFile {
    private static final String TAG = EaseChatRowVideo.class.getSimpleName();

    private ImageView imageView;
    private TextView sizeView;
    private TextView timeLengthView;
    private ImageView playView;

    public EaseChatRowVideo(Context context, boolean isSender) {
        super(context, isSender);
    }

    public EaseChatRowVideo(Context context, ChatMessage message, int position, Object adapter) {
        super(context, message, position, adapter);
    }

	@Override
	protected void onInflateView() {
		inflater.inflate(!showSenderType ? R.layout.ease_row_received_video
                : R.layout.ease_row_sent_video, this);
	}

	@Override
	protected void onFindViewById() {
	    imageView = ((ImageView) findViewById(R.id.chatting_content_iv));
        sizeView = (TextView) findViewById(R.id.chatting_size_iv);
        timeLengthView = (TextView) findViewById(R.id.chatting_length_iv);
        playView = (ImageView) findViewById(R.id.chatting_status_btn);
        percentageView = (TextView) findViewById(R.id.percentage);
	}

	@Override
	protected void onSetUpView() {
        if(bubbleLayout != null) {
            bubbleLayout.setBackground(null);
        }
        VideoMessageBody videoBody = (VideoMessageBody) message.getBody();

        if (videoBody.getDuration() >= 0) {
            String time;
            if(videoBody.getDuration() > 1000) {
                time = EaseDateUtils.toTime(videoBody.getDuration());
            }else {
                time = EaseDateUtils.toTimeBySecond(videoBody.getDuration());
            }
            timeLengthView.setText(time);
        }

        if (message.direct() == ChatMessage.Direct.RECEIVE) {
            if (videoBody.getVideoFileLength() > 0) {
                String size = TextFormater.getDataSize(videoBody.getVideoFileLength());
                sizeView.setText(size);
            }
        } else {
            long videoFileLength = videoBody.getVideoFileLength();
            sizeView.setText(TextFormater.getDataSize(videoFileLength));
        }

        EMLog.d(TAG,  "video thumbnailStatus:" + videoBody.thumbnailDownloadStatus());
        if (message.direct() == ChatMessage.Direct.RECEIVE) {
            if (videoBody.thumbnailDownloadStatus() == FileMessageBody.EMDownloadStatus.DOWNLOADING) {
                imageView.setImageResource(R.drawable.ease_default_image);
            } else {
                // System.err.println("!!!! not back receive, show image directly");
                imageView.setImageResource(R.drawable.ease_default_image);
                showVideoThumbView(message);
            }
        }else{
            if (videoBody.thumbnailDownloadStatus() == FileMessageBody.EMDownloadStatus.DOWNLOADING ||
                    videoBody.thumbnailDownloadStatus() == FileMessageBody.EMDownloadStatus.PENDING ||
                    videoBody.thumbnailDownloadStatus() == FileMessageBody.EMDownloadStatus.FAILED) {
                if(progressBar != null) {
                    progressBar.setVisibility(View.INVISIBLE);
                }
                if(percentageView != null) {
                    percentageView.setVisibility(View.INVISIBLE);
                }
                if(videoBody.thumbnailDownloadStatus() == FileMessageBody.EMDownloadStatus.PENDING) {
                    showVideoThumbView(message);
                }else {
                    imageView.setImageResource(R.drawable.ease_default_image);
                }
            } else {
                if(progressBar != null) {
                    progressBar.setVisibility(View.GONE);
                }
                if(percentageView != null) {
                    percentageView.setVisibility(View.GONE);
                }
                imageView.setImageResource(R.drawable.ease_default_image);
                showVideoThumbView(message);
            }
        }

        setImageIncludeThread(imageView);
	}

    /**
     * show video thumbnails
     * @param message
     */
    @SuppressLint("StaticFieldLeak")
    private void showVideoThumbView(final ChatMessage message) {
        ViewGroup.LayoutParams params = EaseImageUtils.showVideoThumb(context, imageView, message);
        //setBubbleView(params.width, params.height);
    }

    private void setBubbleView(int width, int height) {
        ViewGroup.LayoutParams params = bubbleLayout.getLayoutParams();
        params.width = width;
        params.height = height;
    }

}

File: ease-im-kit/src/main/kotlin/io/agora/uikit/widget/chatrow/EaseChatRowVideo.kt
package io.agora.uikit.widget.chatrow

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.widget.ImageView
import android.widget.TextView
import io.agora.uikit.R
import io.agora.uikit.common.ChatDownloadStatus
import io.agora.uikit.common.ChatLog
import io.agora.uikit.common.ChatMessage
import io.agora.uikit.common.ChatMessageStatus
import io.agora.uikit.common.ChatTextFormater
import io.agora.uikit.common.ChatVideoMessageBody
import io.agora.uikit.common.extensions.isSend
import io.agora.uikit.common.extensions.isSuccess
import io.agora.uikit.common.extensions.loadImageFromMessage
import io.agora.uikit.common.helper.DateFormatHelper
import io.agora.uikit.common.utils.EaseFileUtils

open class EaseChatRowVideo @JvmOverloads constructor(
    private val context: Context,
    private val attrs: AttributeSet? = null,
    private val defStyleAttr: Int = 0,
    isSender: Boolean = false
) : EaseChatRowFile(context, attrs, defStyleAttr, isSender) {
    private val imageView: ImageView? by lazy { findViewById(R.id.chatting_content_iv) }
    private val sizeView: TextView? by lazy { findViewById(R.id.chatting_size_iv) }
    private val timeLengthView: TextView? by lazy { findViewById(R.id.chatting_length_iv) }
    private val playView: ImageView? by lazy { findViewById(R.id.chatting_status_btn) }

    override fun onInflateView() {
        inflater.inflate(
            if (!isSender) R.layout.ease_row_received_video else R.layout.ease_row_sent_video,
            this
        )
    }

    override fun onSetUpView() {
        bubbleLayout?.background = null
        playView?.visibility = GONE
        message?.run {
            val videoBody = body as ChatVideoMessageBody
            if (isSuccess() && videoBody.thumbnailDownloadStatus() == ChatDownloadStatus.DOWNLOADING) {
                setMessageDownloadCallback(true)
            }
            if (videoBody.duration >= 0) {
                val time = if (videoBody.duration > 1000) {
                    DateFormatHelper.toTime(videoBody.duration)
                } else {
                    DateFormatHelper.toTimeBySecond(videoBody.duration)
                }
                timeLengthView?.text = time
            }
            if (!isSend()) {
                if (videoBody.videoFileLength > 0) {
                    val size: String = ChatTextFormater.getDataSize(videoBody.videoFileLength)
                    sizeView?.text = size
                }
            } else {
                val videoFileLength: Long = videoBody.videoFileLength
                sizeView?.text = ChatTextFormater.getDataSize(videoFileLength)
            }
            ChatLog.d(TAG, "video thumbnailStatus:" + videoBody.thumbnailDownloadStatus())
            if (!isSend()) {
                if (videoBody.thumbnailDownloadStatus() === ChatDownloadStatus.DOWNLOADING) {
                    //
                } else {
                    // System.err.println("!!!! not back receive, show image directly");
                    showVideoThumbView(this)
                }
            } else {
                if (status() == ChatMessageStatus.SUCCESS) {
                    if (videoBody.thumbnailDownloadStatus() === ChatDownloadStatus.DOWNLOADING
                        || videoBody.thumbnailDownloadStatus() === ChatDownloadStatus.PENDING
                        || videoBody.thumbnailDownloadStatus() === ChatDownloadStatus.FAILED) {
                        progressBar?.visibility = INVISIBLE
                        if (videoBody.thumbnailDownloadStatus() === ChatDownloadStatus.PENDING) {
                            showVideoThumbView(this)
                        } else if (videoBody.thumbnailDownloadStatus() === ChatDownloadStatus.FAILED){
                            imageView?.setImageResource(R.drawable.ease_default_video_thumbnail)
                        }
                    } else {
                        progressBar?.visibility = GONE
                        showVideoThumbView(this)
                    }
                } else {
                    showVideoThumbView(this)
                }
            }
            // If local thumb exists, show video icon directly
            if (EaseFileUtils.isFileExistByUri(context, videoBody.localThumbUri)) {
                playView?.visibility = VISIBLE
            }
            //setImageIncludeThread(imageView)
        }
    }

    override fun onDownloadAttachmentSuccess() {
        super.onDownloadAttachmentSuccess()
        progressBar?.visibility = GONE
        playView?.visibility = VISIBLE
        message?.run {
            showVideoThumbView(this)
        }
    }

    override fun onDownloadAttachmentProgress(progress: Int) {
        super.onDownloadAttachmentProgress(progress)
        progressBar?.visibility = visibility
    }

    override fun onDownloadAttachmentError(code: Int, error: String?) {
        super.onDownloadAttachmentError(code, error)
        progressBar?.visibility = GONE
        playView?.visibility = GONE
        imageView?.setImageResource(R.drawable.ease_default_video_thumbnail)
    }

    /**
     * show video thumbnails
     * @param message
     */
    @SuppressLint("StaticFieldLeak")
    private fun showVideoThumbView(message: ChatMessage) {
        imageView?.loadImageFromMessage(message, onSuccess = {
            playView?.visibility = VISIBLE
        }, onError = {
            playView?.visibility = GONE
        })
    }

    companion object {
        private val TAG = EaseChatRowVideo::class.java.simpleName
    }
}